### DISCUSSION TITLE
`Bun.watch` - fast and reliable file watching api

### DISCUSSION THREAD

**Discussion by vjpr** [oven-sh/bun#1571] [Ideas]
File watching is very common.

For node.js, `chokidar` is the most popular.

There is also Facebook's `watchman` which runs a daemon that ensures at most one callback is setup per directory. It is used by default if available in Jest.

Would be nice to have a good reliable api for file watching. There are a lot of issues with `fs.watch`.

It would be really nice to have an implementation that could work well with a persistent watcher (like `watchman` does).

# Related

`fs.watch` is not implemented yet: https://github.com/oven-sh/bun/issues/832

**Comment by Jarred-Sumner** [comment]
We will add a builtin watch API 

Haven't gotten to it yet

**Comment by tuyen-at-work** [comment]
@vjpr @Jarred-Sumner Looks like it is done now?
https://bun.sh/guides/read-file/watch

**Reply by bjesuiter** [reply]
I'm not sure.
This seems not to emit all events when they come in rapid succession.

**Reply by 7flash** [reply]
> I'm not sure. This seems not to emit all events when they come in rapid succession.

really?? how many events you emit to manifest this behaviour?

**Reply by guest271314** [reply]
Looks like only 1 event is dispatched. When the name changes.

**Reply by bjesuiter** [reply]
> > I'm not sure. This seems not to emit all events when they come in rapid succession.
> 
> really?? how many events you emit to manifest this behaviour?

I'm watching the Build Output of ng Build --Watch, so nothing too crazy. 
at most, there are around 350 file changes after the build. 

Since the angular cli emits files relatively random, I buffer all file change events and emit them after 1 second of no change to handle all file changes after a build at once. 
And the bun watcher did emit less events than chokidar. 

Maybe @guest271314 is right and it only emitted "rename" events. I don't remember, it's too long ago that I tried it.
